### DISCUSSION TITLE
Exception when isolation repeatable_read in postgresql

### DISCUSSION THREAD

**Discussion by chepaksov** [vladimir-bukhtoyarov/bucket4j#525] [Q&A] [answered]
Hi.
I use bucket4j with advisory pg.
The app has global isolation repeatable read.
I see exception 
org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "rate_limiter_bucket_pkey"
      Detail: Key (id)=(up) already exists., org.postgresql.util.PSQLException: ERROR: could not serialize access due to concurrent update]

When I restore read commited, then ok.
I can't change global isolation.
How I fix problem ?

**Comment by vladimir-bukhtoyarov** [comment]
Hello @chepaksov,

I suppose that you use [PostgreSQLSelectForUpdateBasedProxyManager](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-postgresql/src/main/java/io/github/bucket4j/postgresql/PostgreSQLSelectForUpdateBasedProxyManager.java)). Try switch to [PostgreSQLadvisoryLockBasedProxyManager](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-postgresql/src/main/java/io/github/bucket4j/postgresql/PostgreSQLadvisoryLockBasedProxyManager.java), it should help. See documentation there https://bucket4j.com/8.14.0/toc.html#postgresqladvisorylockbasedproxymanager

**Comment by chepaksov** [comment]
Ok.
I already use PostgreSQLadvisoryLockBasedProxyManager.
CREATE TABLE IF NOT EXISTS rate_limiter_bucket(id varchar PRIMARY KEY, state BYTEA, expires_at BIGINT, explicit_lock BIGINT);
![image](https://github.com/user-attachments/assets/30d83934-0572-48b9-b4c6-8d9a2255adb2)
![image](https://github.com/user-attachments/assets/0b8d67a6-e8ad-4924-900a-6d2e3f45742d)
![image](https://github.com/user-attachments/assets/f2cb976c-3a5a-4765-b917-f956779f8319)

DB provider zonky

ALTER DATABASE xellfclijtkb SET DEFAULT_TRANSACTION_ISOLATION TO 'repeatable read' ;

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Wow, it is surprisly. I have no another ideas that will be helpful to share common datasource. Just create dedicated datasource for bucket4j and set "read commited" isolation level for this data-source.

**Comment by chepaksov** [comment]
Thanks.
It resolved this problem.
